extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts

```diff
@@ -0,0 +1,157 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as vscode from 'vscode';
+import * as azdata from 'azdata';
+import * as nls from 'vscode-nls';
+import { IEndPoint, IControllerError, getEndPoints } from '../controller/clusterControllerApi';
+import { ControllerTreeDataProvider } from '../tree/controllerTreeDataProvider';
+import { TreeNode } from '../tree/treeNode';
+import { showErrorMessage } from '../utils';
+
+const localize = nls.loadMessageBundle();
+
+export class AddControllerDialogModel {
+	constructor(
+		public treeDataProvider: ControllerTreeDataProvider,
+		public node?: TreeNode,
+		public prefilledUrl?: string,
+		public prefilledUsername?: string,
+		public prefilledPassword?: string,
+		public prefilledRememberPassword?: boolean
+	) {
+		this.prefilledUrl = prefilledUrl || (node && node['url']);
+		this.prefilledUsername = prefilledUsername || (node && node['username']);
+		this.prefilledPassword = prefilledPassword || (node && node['password']);
+		this.prefilledRememberPassword = prefilledRememberPassword || (node && node['rememberPassword']);
+	}
+
+	public async onComplete(url: string, username: string, password: string, rememberPassword: boolean): Promise<void> {
+		let response = await getEndPoints(url, username, password, true);
+		if (response && response.request) {
```

Comment:
What happens if there's no response? If this is null or missing key values, what should happen?

Reply:
it is sending "endpoints" request to server, and the API client always provides response or error, and error cases such as auth failure or internal server...are covered by catch clause.
No response without failure cannot happen.

Reply:
If that's the case, we can remove the if check? Basically if we keep this if clause, we should handle the 'else' case. If we never expect it to happen, let's remove the check